AnimationCurve

shrinktofit opened the pull request:
Re: cocos-creator/3d-tasks#

Changelog:
 * 

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

shrinktofit commented:
This PR tends to be in version 3.2+. Please do not merge it into 3.1. @holycanvas

shrinktofit commented:
The concepts and design document can be found at [here](https://github.com/cocos-creator/3d-tasks/issues/2618#issuecomment-819347989).

shrinktofit committed:
AnimationCurve

shrinktofit committed:
Mark curve frame values as uniquelyReferenced

shrinktofit committed:
Adjust modules

shrinktofit committed:
Update

shrinktofit committed:
Manually implements KeyframeCurve.iterator

shrinktofit committed:
TargetPath

shrinktofit committed:
Update

shrinktofit committed:
Fix BUG

shrinktofit committed:
Fix

shrinktofit committed:
Add size track

shrinktofit committed:
Update

shrinktofit committed:
Fix size; unit tests

gameall3d committed:
fix split animation (#24)

shrinktofit committed:
CCONB animation clip; Remove integer track/channel/curve

shrinktofit committed:
Optimize key shared curves

shrinktofit committed:
Remove keyframecurve.empty

shrinktofit committed:
Exotic animation

shrinktofit committed:
Fix unit tests

shrinktofit committed:
start/end -> left/right

shrinktofit committed:
Curve range rename

shrinktofit committed:
Easing methods

shrinktofit committed:
Fix untyped track

shrinktofit committed:
Fix

gameall3d committed:
fix upgradUnTypedTracks (#26)

gameall3d reviewed (approved):
LGTM

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,353 @@ export class AnimationClip extends Asset {
+        const ratios: number[] = [];
+        const eventGroups: IAnimationEventGroup[] = [];
+        const events = this.events.sort((a, b) => a.frame - b.frame);
+        for (const eventData of events) {
```
不要使用 for of

pandamicro replied:
请检查所有的使用

pandamicro replied:
遗漏

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,353 @@ export class AnimationClip extends Asset {
+        this._runtimeEvents = {
+            ratios,
+            eventGroups,
+        };
```
感受到垃圾的产生

shrinktofit replied:
这里是必须创建对象的，可以看看 runtimeEvent 的定义；这个方法只会在设置时候调用（一般是编辑器）

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,353 @@ export class AnimationClip extends Asset {
+            result[joint] = new BoneGlobalTransform();
+            return result;
+        }, {} as Record<string, BoneGlobalTransform>);
+        for (const joint of Object.keys(skeletonFrames)) {
```
for of Object.keys 使用 for in 替代

pandamicro replied:
遗漏

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,353 @@ export class AnimationClip extends Asset {
+                const parentJointName = joint.substring(0, parentJoint);
+                const parentJointFrame = skeletonFrames[parentJointName];
+                if (!parentJointFrame) {
+                    warn(`Seems like we have animation for ${joint} but are missing its parent joint ${parentJointName} in animation?`);
```
使用 warnID，检查所有 warn, log, error 的使用

pandamicro replied:
遗漏

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,353 @@ export class AnimationClip extends Asset {
+    /**
+     * @zh 曲线可引用的所有时间轴。
+     * @en Frame keys referenced by curves.
+     * @deprecated Since V3.1.
```
deprecated 信息需要给出替代方式的指引

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -354,91 +514,655 @@ export class AnimationClip extends Asset {
+     * DO NOT use it in your code since it might be removed for the future at any time.
+     * @deprecated Since V3.1.
+     */
+    public syncLegacyData () {
```
所有函数应该在私有属性之后

pandamicro commented on cocos/core/animation/animation-state.ts:
```diff
@@ -367,22 +281,25 @@ export class AnimationState extends Playable {
+    /**
+     * For internal usage. Really hack...
+     */
+    protected _doNotCreateEval = false;
```
AnimationClip 里面的是应该存在的吗？有重复吗？

pandamicro commented on cocos/core/animation/exotic-animation/exotic-animation.ts:
```diff
@@ -0,0 +1,543 @@
+ * - currently only generated imported from model file.
+ */
+@ccclass(`${CLASS_NAME_PREFIX_ANIM}ExoticAnimation`)
+export class ExoticAnimation {
```
这个命名有其他引擎的先例吗？如果只是模型动画，能不能用 ModelAnimation 或者 ImportedAnimation

pandamicro commented on cocos/core/animation/exotic-animation/exotic-animation.ts:
```diff
@@ -0,0 +1,543 @@
+        const {
+            _values: values,
+            _isQuantized: isQuantized,
+        } = this;
```
这是什么骚操作？

shrinktofit replied:
等价于 const a = this._a

pandamicro commented on cocos/core/animation/tracks/color-track.ts:
```diff
@@ -0,0 +1,71 @@
+        }
+        if (this._w) {
+            this._result.a = this._w.evaluate(time);
+        }
```
感觉会很慢

pandamicro commented on cocos/core/animation/tracks/quat-track.ts:
```diff
@@ -0,0 +1,29 @@
+export class QuatTrackEval {
+    constructor (private _curve: QuaternionCurve) {
+
+    }
```
为什么保留空构造

shrinktofit replied:
这里是一种 TypeScript 技巧，`constructor(private _a: number)` 等于同时声明了有个成员 `_a`

pandamicro commented on cocos/core/animation/value-proxy-factories/morph-weights.ts:
```diff
@@ -32,6 +32,37 @@ import { ccclass, serializable } from 'cc.decorator';
+ * @zh
+ * 用于设置模型组件目标上指定子网格的指定形状的形变权重的曲线值代理工厂。
+ */
+ @ccclass('cc.animation.MorphWeightValueProxy')
```
整个文件的代码缩进有问题

pandamicro commented on cocos/core/curves/curve.ts:
```diff
@@ -0,0 +1,724 @@
+        leftTangent,
+        leftTangentWeight,
+        easingMethod,
+    }:  Partial<RealKeyframeValue> = { }) {
```
这种构造函数会导致大量的对象创建，也就是垃圾

pandamicro commented on cocos/core/curves/curve.ts:
```diff
@@ -0,0 +1,724 @@
+     * Interpolation method used for this keyframe.
+     */
+    @serializable
+    public interpMode = RealInterpMode.LINEAR;
```
不要用缩写

pandamicro commented on cocos/core/curves/curve.ts:
```diff
@@ -0,0 +1,724 @@
+
+@ccclass('cc.RealKeyframeValue')
+@uniquelyReferenced
+export class RealKeyframeValue {
```
为什么要用 Real 前缀

shrinktofit replied:
因为还有 QuatKeyframeValue

pandamicro commented on cocos/core/curves/curve.ts:
```diff
@@ -0,0 +1,724 @@
+
+    set preExtrap (value) {
+        this._preExtrap = value;
+    }
```
这种直接量属性为什么要加 getter setter

pandamicro commented on cocos/core/curves/keys-shared-curves.ts:
```diff
@@ -0,0 +1,298 @@
+        }
+    }
+
+    protected calculateLocation (time: number, out: TimeLocation) {
```
locate

pandamicro commented on cocos/core/curves/real-curve-param.ts:
```diff
@@ -0,0 +1,60 @@
+/**
+ * The method used for interpolation between values of a keyframe and its next keyframe.
+ */
+export enum RealInterpMode {
```
如果其他引擎找到范例，可以考虑使用 Interp 替代 Interpolation

shrinktofit replied:
已替换

pandamicro commented on cocos/core/geometry/curve.ts:
```diff
@@ -123,38 +120,99 @@ export class AnimationCurve {
+                value: legacyCurve.value,
+                leftTangent: legacyCurve.inTangent,
+                rightTangent: legacyCurve.outTangent,
+            }),
```
大量的垃圾对象

pandamicro reviewed (commented):
- MeshRenderer weight 相关的修改是关联的吗？
- 此次重构有没有潜在的性能问题或者预期的性能提升？
- 内存控制问题可能比较严重，需要测试复杂动画和动画切换，检查垃圾创建的情况
- 能不能让兼容 legacy 数据的代码在不需要的情况下被自动或手动剔除？
- 不仅代码量增加，小文件也很多，需要一个小游戏端包体增加的数据

这和个 PR 被合并的前提还是要在 3.3 阶段能够保障以下几点
- 保障动画模块有性能提升
- 避免运行时大量垃圾对象，导致频繁的 GC
- 保障数据迁移的可靠性
- 包体增加有限

shrinktofit committed:
Rename

.interpMode -> .interpolationMode
RealInterpMode -> RealInterpolationMode
QuaternionInterpMode -> QuaternionInterpolationMode
ExtrapMode -> ExtrapolationMode
.preExtrap -> .preExtrapolation
.postExtrap -> .postExtrapolation
QuaternionCurve -> QuatCurve
QuaternionKeyframeValue -> QuatKeyframeValue

shrinktofit committed:
Update pre/post extrap

shrinktofit committed:
Optimize-1

shrinktofit committed:
Optimize-2

shrinktofit commented:
# Performance Report

- BunnyFrameTest: 20000 cases
- BunnyTransformTest: 20000 cases
- ModelAnimation: 1000 cases

Table cells are counted in FPS

## Desktop Browser

Environment: Microsoft Edge 91.0.864.70, Windows 10 x64

|  | BunnyFrameTest | BunnyTransformTest | ModelAnimation |
| -- | --| --| -- |
| 3.1.2| 54-58 | 16-17 | 31-36 |
| 3.3(without this PR)| 48-52 |14-15 | 33-38 |
| 3.3(including this PR)| | | |

## Mobile Browser

Environment: iOS 14.3

|  | BunnyFrameTest | BunnyTransformTest | ModelAnimation |
| -- | --| --|--|
| 3.1.2| 31-33 and reduce to 28-30 | 12-13 | 20-21|
| 3.3(without this PR)| 29-31 | 12-13 | 21-22 |
| 3.3(including this PR)| 27-28 | 17-18 | 24-25|

shrinktofit commented:
# Package Size Report

With features: core and gfx-webgl1, uglified

Before: 1, 265KB

After: 1, 311KB

shrinktofit committed:
Eliminate for-of

pandamicro commented on cocos/core/animation/animation-clip.ts:
```diff
@@ -198,125 +143,356 @@ export class AnimationClip extends Asset {
-    private _keys: number[][] = [];
+    set duration (value) {
+        this._duration = value;
+    }
```
必要性在哪里？

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            quatTrackEvalStatuses,
+        } = compressedDataEvalStatus;
+
+        for (const curves of this._curves) {
```
for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            });
+        }
+
+        for (const track of this._tracks) {
```
for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            });
+        }
+
+        for (const curves of this._quatCurves) {
```
for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            });
+        }
+
+        for (const track of this._quatTracks) {
```
for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            curves.evaluate(time, result);
+        }
+
+        for (const { type, target, immediate, curves } of compressedTrackEvalStatuses) {
```
for of and  for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            target.setValue(value);
+        }
+
+        for (const { curves, result } of keysSharedQuatCurvesEvalStatues) {
```
for of

pandamicro commented on cocos/core/animation/compression/compressed-data.ts:
```diff
@@ -0,0 +1,317 @@
+            curves.evaluate(time, result);
+        }
+
+        for (const { target, curve } of quatTrackEvalStatuses) {
```
for of

pandamicro commented on cocos/core/animation/legacy-clip-data.ts:
```diff
@@ -0,0 +1,766 @@
+//         trackConstructor: ColorTrack,
+//         properties: [['r', 0], ['g', 1], ['b', 2], ['a', 3]],
+//     } as ConvertMap<Color, ColorTrack>,
+// ];
```
不需要的话就删除

pandamicro commented on cocos/core/animation/legacy-clip-data.ts:
```diff
@@ -0,0 +1,766 @@
+
+    set keys (value) {
+        this._keys = value;
+    }
```
无意义的 getter setter

pandamicro commented on cocos/core/animation/legacy-clip-data.ts:
```diff
@@ -0,0 +1,766 @@
+        return this._commonTargets;
+    }
+
+    set commonTargets (value) {
```
无意义的 getter setter

pandamicro commented on cocos/core/animation/legacy-clip-data.ts:
```diff
@@ -0,0 +1,766 @@
+            return track;
+        });
+
+        for (const legacyCurve of legacyCurves) {
```
for of

shrinktofit committed:
Mark deprecateds

shrinktofit committed:
Hide RealKeyframeValue/QuatKeyframeValue Constructor

shrinktofit committed:
Fix quat keyframe value name

shrinktofit committed:
Error ID

shrinktofit committed:
Fix RealKeyframeValue editor extras

shrinktofit committed:
Elimanate for-of forEach
